tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/plugin/MetricsPluginTest.java

```diff
@@ -198,6 +199,36 @@ public void testForBasicMetricsPresenceAndValue() throws Exception {
       checkMetric(metrics, "artemis.consumer.count", "queue", queueName, 0.0);
    }
 
+   @Test
+   public void testMessageCountWithPaging() throws Exception {
+      final String data = "Simple Text " + UUID.randomUUID().toString();
+      final String queueName = "simpleQueue";
+      final String addressName = "simpleAddress";
+
+      server.getAddressSettingsRepository().getMatch(addressName).setAddressFullMessagePolicy(AddressFullMessagePolicy.PAGE).setMaxSizeBytes(1024 * 10).setPageSizeBytes(1024 * 5);
+
+      session.createQueue(new QueueConfiguration(queueName).setAddress(addressName).setRoutingType(RoutingType.ANYCAST));
+      ClientProducer producer = session.createProducer(addressName);
+      ClientMessage message = session.createMessage(true);
+      message.getBodyBuffer().writeString(data);
+      int messageCount = 0;
+      while (!server.getPagingManager().getPageStore(new SimpleString(addressName)).isPaging()) {
+         producer.send(message);
+         messageCount++;
+      }
+
+      assertEquals(messageCount, server.locateQueue(queueName).getMessageCount());
+      checkMetric(getMetrics(), "artemis.message.count", "queue", queueName, Double.valueOf(messageCount));
+
+      for (int i = 0; i < messageCount; i++) {
+         producer.send(message);
+      }
+      producer.close();
+
+      assertEquals(messageCount * 2, server.locateQueue(queueName).getMessageCount());
```

Comment:
This test fails... run it in loop.

Wait.assertEquals please

Reply:
change this line to

```java
Wait.assertEquals(messageCount, server.locateQueue(queueName)::getMessageCount);
```

Reply:
Done. I ran the test 500 times with no failures.